### PR TITLE
[collapsible][accordion] Add `keepMounted` prop

### PR DESCRIPTION
**Pull request by mj12albert**
Closes https://github.com/mui/base-ui/issues/728

We decided to default `keepMounted` to `false` for consistency with other components.

Since `hiddenUntilFound` requires `keepMounted={true}` to work, using it will override `keepMounted` and show a warning in dev mode if both `hiddenUntilFound` and `keepMounted={false}` are specified.

### CODE REVIEW

**Comment by mj12albert** [comment]
@colmtuite @vladmoroz Do you think it's an issue that 2 props are required to use hidden-until-found now?
```jsx
<Collapsible.Panel hiddenUntilFound keepMounted>
  {/* content */}
</Collapsible.Panel>
```

**Comment by vladmoroz** [comment]
@mj12albert I think we should always mount the panel when `hiddenUntilFound` is used, regardless of `keepMounted`

**Comment by michaldudak** [comment]
If `keepMounted` is undefined, then yes, we can make `hiddenUntilFound` make the component always mounted. But when it's explicitly set to `false`, I'd show a warning (in dev mode only, though).

File: packages/mui-base/src/Collapsible/Panel/CollapsiblePanel.test.tsx [hunk 1]
```diff
@@ -37,4 +49,86 @@ describe('<Collapsible.Panel />', () => {
...
+        panel?.dispatchEvent(event);
+      });
+
+      expect(handleOpenChange.callCount).to.equal(1);
+      expect(panel).to.have.attribute('data-open');
+    });
+
+    // toWarnDev doesn't work reliably with async rendering. To re-enable after it's fixed in the test-utils.
+    // eslint-disable-next-line mocha/no-skipped-tests
+    it.skip('warns when setting both `hiddenUntilFound` and `keepMounted={false}`', async function test() {
```
Comment (mj12albert) [inline packages/mui-base/src/Collapsible/Panel/CollapsiblePanel.test.tsx:118]: I tested this warning manually 😓

**Comment by mj12albert** [comment]
Updated ~ would appreciate a review @vladmoroz @michaldudak 🙏

**Comment by mj12albert** [comment]
Also includes a lot of rework to the panel internals as the Select PR changed the timing of the `useTransitionStatus` hook and set `data-entering` 1 animation frame sooner than before: https://github.com/atomiks/base-ui/commit/cfd185a36784341e7165d17c5fec927fc710e4a8#diff-ecb73c88ec0196a8b56a56bb2b147a4d57e6f461c854c51e7a3109408336fb6c

This broke CSS transition handling, as it was using that 1 frame between the open state changing and `data-entering` to measure the fully expanded size of the panel contents before applying the transition (which starts when `data-entering` is applied)

**Comment by mj12albert** [comment]
Continued in https://github.com/mui/base-ui/pull/889

Requires https://github.com/mui/base-ui/pull/887

**Pull request closed without merging**